### DISCUSSION TITLE
Connection established successfully, but no output is shown on the window

### DISCUSSION THREAD

**Discussion by xavizardKnight** [MarcusWichelmann/MarcusW.VncClient#21] [Q&A] [answered]
Hello! I want to use this library for an Avalonia project which I started working on these lasts days just for fun for a custom viewer for my QEMU VMs, but I cannot make it work with my projects.

I've downloaded this repo and the sample project connects perfectly to my machine (either by 127.0.0.1 or 0.0.0.0; both with ports 5901).

![Captura de pantalla a 2023-05-23 00-31-58](https://github.com/MarcusWichelmann/MarcusW.VncClient/assets/35200790/ba68c51c-22fe-449e-b789-3702b18f7434)

The logs don't show anything suspicious either:
> /home/xavier/RiderProjects/MarcusW.VncClient/samples/AvaloniaVncClient/bin/Debug/netcoreapp3.1/AvaloniaVncClient
info: AvaloniaVncClient.Views.MainWindow[0]
      Hello World!
info: MarcusW.VncClient.RfbConnection[0]
      Connecting to VNC-Server on vnc://127.0.0.1:5901...
warn: MarcusW.VncClient.Protocol.Implementation.DefaultImplementation[0]
      The TurboJPEG library is not installed on the current system. Please expect a performance drop, because the Tight encoding type is not available.
info: MarcusW.VncClient.RfbConnection[0]
      Connection to vnc://127.0.0.1:5901 established successfully.

But when I try to add it to my projects, nothing appears on the screen.
I've created a small project with nothing but an empty window with the VncView component in it, and I followed the README.md steps. 

MainWindow.axaml:
```html
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:fakevm="using:EmptyVNC.Views"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:vnc="clr-namespace:MarcusW.VncClient.Avalonia;assembly=MarcusW.VncClient.Avalonia"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="EmptyVNC.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="EmptyVNC">

    <Design.DataContext><fakevm:MainWindow/></Design.DataContext>

	<vnc:VncView Connection="{Binding RfbConnection}"/>
</Window>
``` 

MainWindow.axaml.cs:
```c#
using System;
using System.Threading;
using System.Threading.Tasks;
using Avalonia.Controls;
using MarcusW.VncClient;
using MarcusW.VncClient.Avalonia.Adapters.Logging;
using MarcusW.VncClient.Protocol.Implementation.Services.Transports;
using MarcusW.VncClient.Protocol.SecurityTypes;
using MarcusW.VncClient.Security;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Hosting;

namespace EmptyVNC.Views
{
	public partial class MainWindow : Window
	{
		public object? RfbConnection { get; set; }

		public MainWindow()
		{
			InitializeComponent();
			
			Connect();
		}

		public async Task Connect()
		{
			IHost host = Host.CreateDefaultBuilder().ConfigureServices((context, services) =>
			{
				services.AddLogging(loggingBuilder =>
				{
					loggingBuilder.AddConsole();
					loggingBuilder.AddProvider(new AvaloniaLoggerProvider());
				});
			}).Build();

			ILogger<MainWindow> logger = host.Services.GetRequiredService<ILogger<MainWindow>>();
			logger.LogInformation("Hello World!");
			
			ILoggerFactory loggerFactory = host.Services.GetRequiredService<ILoggerFactory>();
			VncClient vncClient = new VncClient(loggerFactory);

			IAuthenticationHandler authenticationHandler = new DemoAuthenticationHandler();
			
			ConnectParameters parameters = new ConnectParameters
			{
				TransportParameters = new TcpTransportParameters
				{
					Host = "127.0.0.1",
					Port = 5901
				},
				AuthenticationHandler = authenticationHandler
			};
			
			CancellationToken cancellationToken = CancellationToken.None;

			RfbConnection = await vncClient.ConnectAsync(parameters, cancellationToken).ConfigureAwait(true);
		}
	}
	
	public class DemoAuthenticationHandler : IAuthenticationHandler
	{
		public Task<TInput> ProvideAuthenticationInputAsync<TInput>(RfbConnection connection, ISecurityType securityType, IAuthenticationInputRequest<TInput> request)
			where TInput : class, IAuthenticationInput
		{
			if(typeof(TInput)==typeof(PasswordAuthenticationInput))
			{
				string password = String.Empty;

				return Task.FromResult((TInput)Convert.ChangeType(new PasswordAuthenticationInput(password), typeof(TInput)));
			}

			throw new InvalidOperationException("The authentication input request is not supported by this authentication handler.");
		}
	}
}
```

When I compile the application, the window is empty:
![Captura de pantalla a 2023-05-23 00-31-38](https://github.com/MarcusWichelmann/MarcusW.VncClient/assets/35200790/d3c788e0-90b9-46e6-8290-3368162ff51f)

And in the logs, it appears that the connection gets established successfully, but there's still no output on the screen.

> /home/xavier/RiderProjects/EmptyVNC/bin/Debug/net5.0/EmptyVNC
info: EmptyVNC.Views.MainWindow[0]
      Hello World!
info: MarcusW.VncClient.RfbConnection[0]
      Connecting to VNC-Server on vnc://127.0.0.1:5901...
warn: MarcusW.VncClient.Protocol.Implementation.DefaultImplementation[0]
      The TurboJPEG library is not installed on the current system. Please expect a performance drop, because the Tight encoding type is not available.
info: MarcusW.VncClient.RfbConnection[0]
      Connection to vnc://127.0.0.1:5901 established successfully.

I'm using .NET 5 and Avalonia 0.10; from my Ubuntu MATE 20.04 LTS.
I noticed that the sample project uses .NET Core 3.1, I tried to switch my projects to 3.1 but still the same: white window.

I don't understand why the sample project works, but my projects do not. I've put everything the README says.
If I had to guess, I would say that I need something of MVVM and/or ReactiveUI, but I'm still a noob in these lasts things. I tried adding doing some MVVM tests but for now the same white window appears.

I've tried other VNC clients like Remmina 1.4.2, and it connects perfectly to the machine.

What I'm I missing here?

**Comment by MarcusWichelmann** [comment, marked as answer]
Without trying your code, I could imagine, that your issue is the Binding of the `RfbConnection`.  To use bindings, you should follow the MVVM pattern with a ViewModel and a DataContext set. Read here more about it: https://docs.avaloniaui.net/docs/data-binding/the-datacontext

With your code, I bet the Binding doesn't work and therefore the established `RfbConnection` is never assigned to the `VncView`.

If you don't want to use MVVM, you could also locate the `VncView` control directly from your code-behind by giving it a name, and set the `Connection` property to your `RfbConnection` instance. See here: https://docs.avaloniaui.net/guides/basics/code-behind#locating-controls

I'd highly recommend diving deeper into how Avalonia works.
